### ISSUE TITLE
toplevel_ref_arg should only check functions

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#170]
I'm pretty sure `|ref a| {...}` is valid and the `ref` does something here.

Might be wrong though. cc @llogiq

**Comment by llogiq** [comment]
It is and it does – however it could also be written as `|&a| ...`, if I'm not mistaken.

**Comment by birkenfeld** [comment]
I don't think so - `|&a|` will do the opposite: dereference the reference on call. This can be impossible with non-`Copy` types.

Even for non-closures, I don't think the lint message is accurate; the `ref` is not ignored. `ref a: T` is similar to `a: &T` but you call it with a value that is moved.

**Comment by llogiq** [comment]
OK, now that I have had my coffee: ref means take arg by immutable reference – which can be useful even if art is copy, because the compiler will stop you from mutating it, whereas you could accidentally mutate the copy otherwise.

**Comment by killercup** [comment]
I think I just ran into this. This code ([source](https://github.com/killercup/cargo-edit/blob/ea7dc7a70b67ad09736b91014fe28a7fbc22b521/src/manifest.rs#L40-L44)):

```rust
.and_then(|ref dir| search(dir, file).map_err(From::from)))
```

gives this warning:

```text
src/manifest.rs:43:29: 43:36 warning: `ref` directly on a function argument is ignored. Consider using a reference type instead., #[warn(toplevel_ref_arg)] on by default
src/manifest.rs:43                  .and_then(|ref dir| search(dir, file).map_err(From::from)))
                                               ^~~~~~~
note: in expansion of closure expansion
src/manifest.rs:43:28: 43:75 note: expansion site
note: in expansion of closure expansion
src/manifest.rs:42:14: 43:76 note: expansion site
```

I believe this is perfectly valid, though, and I don't really want to write this instead:

```rust
.and_then(|dir| search(&*dir, file).map_err(From::from))
```

(This compiles with no clippy warnings. Maybe clippy should even suggest using `ref` on the argument instead of `&*` here?)

**Comment by Manishearth** [comment]
Yeah, the `ref` thing should only apply to functions, not closures

**Comment by birkenfeld** [comment]
What is so different about functions? Why is this not useful to write there?

**Comment by Manishearth** [comment]
For functions, pass by ref is done with an `&T` argument.

With closures, you usually don't specify the argument types, so `ref` is useful.
